### ISSUE TITLE
CppMod ImGui interface

### ISSUE THREAD

**Issue by trumank** [UE4SS-RE/RE-UE4SS#110]
CppMods should have a way to add a tab to the GUI for mod specific settings / widgets.

**Comment by UE4SS** [comment]
I think this is a good idea.

Do you have any ideas for how we could deal with exporting ImGui to each C++ mod ?

The problem is that the ImGui context isn't preserved across dll boundaries so if we call an ImGui function from a C++ mod we'll call it on a nullptr ImGui context.

At the moment, the only solution I have is to call `ImGui::SetCurrentContext` & `ImGui::SetAllocatorFunctions` from each C++ mod.
Like this for example:
```c++
ImGui::SetCurrentContext(UE4SSProgram::get_current_imgui_context());
ImGuiMemAllocFunc alloc_func{};
ImGuiMemFreeFunc free_func{};
void* user_data{};
UE4SSProgram::get_current_imgui_allocator_functions(&alloc_func, &free_func, &user_data);
ImGui::SetAllocatorFunctions(alloc_func, free_func, user_data);
```
It would have to be part of some initialization (like `on_program_start`), and executed before any ImGui functions are called.
It's rather bulky and annoying to have to include this in every C++ mod but it might be acceptable (maybe with some refactoring to reduce & simplify) but obviously an alternative to this would be better.

**Comment by trumank** [comment]
Ah, no I was not aware of the ImGui context issue. I imagine the bulk of it could be shoved into an `#include` so it's not necessary to copy and paste the same block.

So far I've been only thinking about how the actual interface would work. The most straightforward (stateless as far as UE4SS is concerned) would be to have two functions, one which returns whether the mod has a GUI (or an optional string with the GUI name) and one which is called to actually render the tab:

```C++
auto CppMod::gui_title() -> std::optional<std::wstring>; // none if no GUI; or wstring if GUI
auto CppMod::render_gui() -> void; // called by UE4SS when it's time to render the GUI
```

The other option would be to explicitly register GUI tabs/windows so it's possible to have many or none at all. The exact interface for this may depend on the ImGui docking branch which I've wanted to explore: https://github.com/ocornut/imgui/issues/2109

**Comment by UE4SS** [comment]
> The other option would be to explicitly register GUI tabs/windows so it's possible to have many or none at all. The exact interface for this may depend on the ImGui docking branch which I've wanted to explore: [ocornut/imgui#2109](https://github.com/ocornut/imgui/issues/2109)

We had someone else start implementing the docking branch but I don't know what the status is of that.
@narknon Do you still have contact with that person ? If so, could you ask them what the status is ?

**Comment by UE4SS** [comment]
> Ah, no I was not aware of the ImGui context issue. I imagine the bulk of it could be shoved into an `#include` so it's not necessary to copy and paste the same block.

I think we can make a macro for the ImGui stuff that we demand is placed inside one of the init functions.
It cannot be in the constructor as that executes before the GUI has been set up by UE4SS.
The `on_program_start` function is safe, and I believe so is `on_unreal_init`.


> So far I've been only thinking about how the actual interface would work. The most straightforward (stateless as far as UE4SS is concerned) would be to have two functions, one which returns whether the mod has a GUI (or an optional string with the GUI name) and one which is called to actually render the tab:
> 
> ```c++
> auto CppMod::gui_title() -> std::optional<std::wstring>; // none if no GUI; or wstring if GUI
> auto CppMod::render_gui() -> void; // called by UE4SS when it's time to render the GUI
> ```

My current implementation is close to your second idea:
```c++
class MyTestTab : public GUI::GUITab
{
public:
    MyTestTab() : RC::GUI::GUITab()
    {
        TabName = "My Test Tab";
    }

    auto render() -> void override
    {
        ImGui::Text("This is my test tab!");
    }
};

class MyAwesomeMod : public RC::CppUserModBase
{
public:
    std::unique_ptr<MyTestTab> m_test_tab{};

    auto on_program_start() -> void override
    {
        UE4SS_ENABLE_IMGUI()
        m_test_tab = std::make_unique<MyTestTab>();
        UE4SSProgram::get_program().add_gui_tab(m_test_tab.get());
    }
};
```
This interface can change of course if needed.

I'm also considering whether to put these tabs in the main tab bar or whether to create a tab that has another tab bar inside it specifically for mod tabs.

**Comment by trumank** [comment]
> My current implementation is close to your second idea:
> *[snip]*
> This interface can change of course if needed.
This looks good. :+1: 

> I'm also considering whether to put these tabs in the main tab bar or whether to create a tab that has another tab bar inside it specifically for mod tabs.
I think adding them to the main tab bar would be fine, at least for now. My current use case isn't for a mod per se, but closer to a plugin: a debugger for the kismet VM.

> Speaking of nothing, we have a docs site here https://github.com/UE4SS-RE/docs. I mention this only because I saw your docs branch and just want to let you know to not bother with it if you intended to make it specifically as a UE4SS main docs site.
I was experimenting with moving the docs into the main repo as I believe that gives them a better chance of remaining updated and in sync with the rest of the codebase. Unfortunately moving repos means the pages URL would change without doing some other shenanigans like pushing the docs out to the docs repo with an action, so I haven't decided if it's ultimately a good idea or not.

**Comment by trumank** [comment]
Implemented by #111
